designs/metrics.md

```diff
@@ -0,0 +1,296 @@
+
+Following Kubernetes best practices, **production Prometheus metrics are co-located with their implementation components**:
+
+1. **Provider-specific metrics** → Live within provider packages (e.g., `pkg/providers/instance/`)
+2. **Controller-specific metrics** → Live within controller packages (e.g., `pkg/controllers/*/`)
+3. **Component-specific metrics** → Live with the component (batcher, cache, etc.)
+
+### Why Co-location?
+
+#### 1. **Locality of Reference**
+- Metrics are defined next to the code that uses them
+- Easy to understand what metrics a component exposes
+- Reduces cognitive load when reading/maintaining code
+
+#### 2. **Package Encapsulation**
+- Each component owns its metrics
+- Changes to component behavior and metrics happen together
+- No cross-package dependencies for metrics
+
+## Specific Changes in Detail
+
+### Label Design
+
+Each component defines **labels relevant to its operations**:
+
+```go
+// Example from instance provider
+ComponentMetric = prometheus.NewCounterVec(
+    prometheus.CounterOpts{
+        Namespace: coremetrics.Namespace,  // From centralized constants
+        Subsystem: componentSubsystem,
+        Name:      "operation_total",
+        Help:      "Total number of operations.",
+    },
+    []string{"label1", "label2"},  // Component-specific labels
+)
+```
+
+**Label Selection Guidelines:**
+- Choose labels that help diagnose component-specific issues
+- Keep cardinality reasonable (avoid high-cardinality labels like instance IDs)
+- Use labels that align with how the component operates
+
+### Label Constants
+
+Defined locally in each component's metrics file:
+
+```go
+const (
+    componentSubsystem = "component_name"
+    label1 = "label_one"
+    label2 = "label_two"
+)
+```
+
+These are **component-specific** and don't need to be shared unless multiple components use identical labels.
+
+### Metric Call Pattern
+
+Using standard Prometheus API:
+
+```go
+// Simple, direct metric recording
+ComponentMetric.WithLabelValues(value1, value2).Inc()
+```
+
+**Benefits:**
+- Standard Prometheus pattern familiar to all Go developers
+- No custom abstraction to learn
+- Direct control over label values
+- Clear at call site what's being measured
+
+## Centralized Metrics Package Role
+
+The `pkg/metrics/` package remains important but serves a **different purpose**:
+
+### Purpose: Infrastructure & Shared Utilities
+
+1. **Constants**
+   ```go
+   const (
+       Namespace = "karpenter_aks"  // Shared across all metrics
+   )
+   ```
+
+2. **Builder Patterns**
+   ```go
+   func NewSubsystemMetrics(subsystem string, labels []string, buckets []float64) *SubsystemMetrics
+   ```
+
+## Migration Path for Other Components
+
+When adding metrics to other components, follow this pattern:
+
+### 1. Create Component Metrics File
+
+```go
+// pkg/providers/<component>/metrics.go
+package component
+
+import (
+    coremetrics "github.com/Azure/karpenter-provider-azure/pkg/metrics"
```

Comment:
minor: This isn't usually what we mean by `core`. Core means upstream (karpenter-core). If it's from karpenter-provider-azure, I don't' think it's `core`.